bin/propolis-server/src/lib/vcpu_tasks.rs

```diff
@@ -0,0 +1,212 @@
+//! Tasks for vCPU backing threads and controls for them.
+
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc,
+};
+
+use propolis::{
+    bhyve_api,
+    exits::{self, VmExitKind},
+    vcpu::Vcpu,
+    VmEntry, VmError,
+};
+use slog::{debug, error, info};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum VcpuTaskError {
+    #[error("Failed to spawn a vCPU backing thread: {0}")]
+    BackingThreadSpawnFailed(std::io::Error),
+
+    #[error("Failed to create a tokio runtime: {0}")]
+    TokioRuntimeCreationFailed(std::io::Error),
+}
+
+pub struct VcpuTasks {
+    tasks: Vec<(propolis::tasks::TaskCtrl, std::thread::JoinHandle<()>)>,
+    generation: Arc<AtomicUsize>,
+    rt: tokio::runtime::Handle,
+}
+
+pub trait VcpuEventHandler: Send + Sync {
+    fn suspend_halt_event(&self, vcpu_id: i32);
+    fn suspend_reset_event(&self, vcpu_id: i32);
+    fn suspend_triple_fault_event(&self, vcpu_id: i32);
+    fn unhandled_vm_exit(&self, vcpu_id: i32, exit: VmExitKind);
+    fn io_error_event(&self, vcpu_id: i32, error: std::io::Error);
+}
+
+impl VcpuTasks {
+    pub fn new(
+        instance: propolis::instance::InstanceGuard,
+        event_handler: Arc<dyn VcpuEventHandler>,
```

Comment:
Fixed in 87968b0. (For posterity, the original reason for this was largely symmetry with what we do for chipset events, which symmetry we will need if we ever move the vCPU task loops down into propolis-lib proper, but since we're not doing that in this PR we can just use the concrete type for now and reimplement this later if need be.)